nuxeo-ai-model/src/main/java/org/nuxeo/ai/bulk/BulkRemoveEnrichmentAction.java

```diff
@@ -0,0 +1,151 @@
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.stream.StreamProcessorTopology;
+
+/**
+ *  A bulk action to remove enrichment metadata
+ */
+public class BulkRemoveEnrichmentAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "bulkEnrichRemove";
+
+    public static final String PARAM_MODEL = "modelId";
+
+    public static final String PARAM_XPATHS = "xpaths";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+
+        return Topology.builder()
+                       .addComputation(RemovalComputation::new,
+                                       Arrays.asList(INPUT_1 + ":" + ACTION_NAME, //
+                                                     OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class RemovalComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(RemovalComputation.class);
+
+        protected String modelId;
+
+        protected Set<String> xPaths;
+
+        public RemovalComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            Serializable modelParam = command.getParam(PARAM_MODEL);
+            Serializable xParam = command.getParam(PARAM_XPATHS);
+            modelId = modelParam != null ? String.valueOf(modelParam) : null;
+            xPaths = xParam != null ? getSet(xParam) : Collections.emptySet();
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            DocMetadataService metadataService = Framework.getService(DocMetadataService.class);
+            for (String id : ids) {
+                try {
+                    DocumentModel doc = session.getDocument(new IdRef(id));
+                    if (doc.hasFacet(ENRICHMENT_FACET)) {
+                        SuggestionMetadataWrapper wrapper = new SuggestionMetadataWrapper(doc);
+                        Set<String> removalProperties;
+                        if (modelId != null) {
+                            List<LabelSuggestion> suggestions = wrapper.getSuggestionsByModel(modelId);
+                            removalProperties = suggestions.stream()
+                                                           .map(LabelSuggestion::getProperty)
+                                                           .collect(Collectors.toSet());
+                        } else if (!xPaths.isEmpty()) {
+                            removalProperties = xPaths;
+                        } else {
+                            removalProperties = wrapper.getAutoProperties();
+                        }
+
+                        for (String xPath : removalProperties) {
```

Comment:
nested loop is the only one thing that bothers me here.

Reply:
The inner loop is just a small Set of property names, probably less than 5 values.